dateutil/parser/isoparser.py

```diff
@@ -0,0 +1,424 @@
+            day = int(dt_str[pos:pos + 2])
+            year = self._default_year
+
+            if month == 2 and day == 29:
+                # Calcualtes the latest leap year
+                year -= year % 4
+                if (year % 400) and not (year % 100):
+                    year -= 4
+
+            return [year, month, day], pos + 2
+
+        # All other uncommon ISO formats start with the year
+        year = int(dt_str[0:4])
+
+        pos = 4 + (dt_str[4:5] == self._DATE_SEP)   # Skip '-' if it's there
+        if dt_str[pos:pos + 1] == b'W':
+            # YYYY-?Www-?D?
+            pos += 1
+            weekno = int(dt_str[pos:pos + 2])
+            pos += 2
+
+            dayno = 1
+            if len(dt_str) > pos:
+                if dt_str[pos:pos + 1] == self._DATE_SEP:
+                    # YYYY-W
+                    if dt_str[4:5] != self._DATE_SEP:
+                        raise ValueError('Inconsistent use of dash separator')
+                    pos += 1
+
+                dayno = int(dt_str[pos:pos + 1])
+                pos += 1
+
+            base_date = self._calculate_weekdate(year, weekno, dayno)
+        else:
+            # YYYYDDD or YYYY-DDD
+            ordinal_day = int(dt_str[pos:pos + 3])
+            pos += 3
+
+            if ordinal_day < 1 or ordinal_day > (365 + calendar.isleap(year)):
+                raise ValueError('Invalid ordinal day' +
+                                 ' {} for year {}'.format(ordinal_day, year))
+
+            base_date = date(year, 1, 1) + timedelta(days=ordinal_day - 1)
+
+        components = [base_date.year, base_date.month, base_date.day]
+        return components, pos
+
+    @classmethod
+    def _calculate_weekdate(cls, year, week, day):
+        """
+        Calculate the day of corresponding to the ISO year-week-day calendar.
+
+        This function is effectively the inverse of
+        :func:`datetime.date.isocalendar`.
+
+        :param year:
+            The year in the ISO calendar
+
+        :param week:
+            The week in the ISO calendar - range is [1, 53]
+
+        :param day:
+            The day in the ISO calendar - range is [1 (MON), 7 (SUN)]
+
+        :return:
+            Returns a :class:`datetime.date`
+        """
+        if not 0 < week < 54:
+            raise ValueError('Invalid week: {}'.format(week))
+
+        if not 0 < day < 8:     # Range is 1-7
+            raise ValueError('Invalid weekday: {}'.format(day))
+
+        # Get week 1 for the specific year:
+        jan_4 = date(year, 1, 4)   # Week 1 always has January 4th in it
+        week_1 = jan_4 - timedelta(days=jan_4.isocalendar()[2] - 1)
+
+        # Now add the specific number of weeks and days to get what we want
+        week_offset = (week - 1) * 7 + (day - 1)
+        return week_1 + timedelta(days=week_offset)
+
+    @classmethod
+    def _parse_isotime(cls, timestr):
+        len_str = len(timestr)
+        components = [0, 0, 0, 0, None]
```

Comment:
I'd advocate using named variables and putting them all into a tuple at the end

Reply:
That was my original thought, but it makes it much more complicated to exit early from the parsing. The three options you have are: 

1. have a `components` list which you index in a `while` loop that you can `break` from (what I've done here)
2. take a mutable `components` object (basically a list again) as an input and `return` when you're done parsing the string (something I'm strongly considering doing, actually, for speed reasons)
3. use some other complicated method like increasingly nested values.

I think the first one is relatively readable and performant, but came up with the second one while working on the Rust implementation, so I thought I'd try that out and see if it looks any better.